File: scripts/loadDataset.js
// Dataset ingestion -> Postgres.
// AOL query log: derive counts by aggregation (COUNT per normalized query).
//   node scripts/loadDataset.js --dir files/aol_data --min-count 2 --out files/aol_agg.tsv
//   node scripts/loadDataset.js --agg-file files/aol_agg.tsv --top 1000000 --min-count 3
//   node scripts/loadDataset.js --synthetic 120000        # no download
// For the full 35M-row aggregation, give Node more heap:
//   node --max-old-space-size=4096 scripts/loadDataset.js --dir files/aol_data --out files/aol_agg.tsv
const fs = require('fs');
const zlib = require('zlib');
const readline = require('readline');
const path = require('path');
const store = require('../backend/store');

const arg = (name) => { const i = process.argv.indexOf(name); return i >= 0 ? process.argv[i + 1] : undefined; };
const has = (name) => process.argv.includes(name);

function openStream(p) {
  const s = fs.createReadStream(p);
  return p.endsWith('.gz') ? s.pipe(zlib.createGunzip()) : s;
}

async function aggregateInto(file, counter) {
  const rl = readline.createInterface({ input: openStream(file), crlfDelay: Infinity });
  let header = null, qi = 1, delim = '\t', n = 0;
  for await (const line of rl) {
    if (header === null) {
      delim = (line.split('\t').length - 1) >= (line.split(',').length - 1) ? '\t' : ',';
      const cols = line.split(delim).map((c) => c.trim().toLowerCase());
      qi = cols.indexOf('query');
      if (qi < 0) qi = cols.length > 1 ? 1 : 0;
      header = cols;
      continue;
    }
    const parts = line.split(delim);
    if (parts.length <= qi) continue;
    const q = parts[qi].trim().toLowerCase();
    if (!q || q === '-') continue;
    counter.set(q, (counter.get(q) || 0) + 1);
    n++;
  }
  console.log(`  ${path.basename(file)}: ${n.toLocaleString()} query rows (delim=${JSON.stringify(delim)}, col=${qi})`);
}

async function aggregatePaths(paths, minCount) {
  const counter = new Map();
  for (const p of paths) await aggregateInto(p, counter);
  const rows = [];
  for (const [q, c] of counter) if (c >= minCount) rows.push([q, c]);
  return rows;
}

function writeTsv(rows, file) {
  const out = fs.createWriteStream(file);
  for (const [q, c] of rows) out.write(`${q}\t${c}\n`);
  out.end();
  return new Promise((res) => out.on('finish', res));
}

async function readAgg(file, top, minCount) {
  const rows = [];
  const rl = readline.createInterface({ input: fs.createReadStream(file), crlfDelay: Infinity });
  for await (const line of rl) {
    const i = line.lastIndexOf('\t');
    if (i < 0) continue;
    const q = line.slice(0, i), c = parseInt(line.slice(i + 1), 10);
    if (q && c >= minCount) rows.push([q, c]);
  }
  if (top) { rows.sort((a, b) => b[1] - a[1]); return rows.slice(0, top); }
  return rows;
}

function synthetic(n) {
  const heads = ['how to', 'best', 'buy', 'cheap', 'free', 'download', 'what is', 'iphone', 'samsung',
    'java', 'python', 'amazon', 'google', 'weather', 'news', 'movie', 'song', 'recipe', 'near me', 'online', 'review'];
  const tails = ['tutorial', 'price', '2026', 'review', 'online', 'near me', 'for sale', 'guide', 'vs', 'app',
    'login', 'meaning', 'today', 'free', 'pro max', 'case', 'charger', 'stock', 'results', 'live', 'download', 'lyrics'];
  const mids = ['', 'best ', 'new ', 'cheap ', 'top ', 'the '];
  const out = new Map();
  let rank = 1, i = 0;
  while (out.size < n) {
    const h = heads[i % heads.length];
    const m = mids[Math.floor(i / heads.length) % mids.length];
    const t = tails[Math.floor(i / (heads.length * mids.length)) % tails.length];
    const suffix = i >= 5000 ? ` ${Math.floor(i / 5000)}` : '';
    const q = `${h} ${m}${t}${suffix}`.trim();
    if (!out.has(q)) { out.set(q, Math.max(1, Math.floor(10_000_000 / Math.pow(rank, 1.1)))); rank++; }
    i++;
  }
  return [...out.entries()];
}

async function main() {
  const file = arg('--file'), dir = arg('--dir'), synth = arg('--synthetic'), aggFile = arg('--agg-file');
  const out = arg('--out'), top = arg('--top') ? parseInt(arg('--top'), 10) : null;
  const minCount = arg('--min-count') ? parseInt(arg('--min-count'), 10) : 1;

  if (!file && !dir && !synth && !aggFile) {
    console.error('provide --file, --dir, --agg-file, or --synthetic');
    process.exit(1);
  }

  console.log('aggregating...');
  let rows;
  if (synth) rows = synthetic(parseInt(synth, 10));
  else if (aggFile) rows = await readAgg(aggFile, top, minCount);
  else {
    let paths;
    if (dir) paths = fs.readdirSync(dir).filter((f) => f.endsWith('.txt.gz') || f.endsWith('.txt')).sort().map((f) => path.join(dir, f));
    else paths = [file];
    rows = await aggregatePaths(paths, minCount);
    if (top) { rows.sort((a, b) => b[1] - a[1]); rows = rows.slice(0, top); }
  }
  console.log(`${rows.length.toLocaleString()} distinct queries to load`);

  if (out) { await writeTsv(rows, out); console.log(`wrote ${rows.length.toLocaleString()} rows to ${out}`); return; }

  store.initPool();
  try {
    await store.initSchema();
    if (!has('--no-truncate')) await store.truncate();
    await store.bulkLoad(rows);
    console.log(`done. rows in DB: ${(await store.countRows()).toLocaleString()}`);
  } finally {
    await store.closePool();
  }
}

main().catch((e) => { console.error(e); process.exit(1); });

File: scripts/load_dataset.py
"""Dataset ingestion -> Postgres.

Primary dataset: AOL search query log (Kaggle: dineshydv/aol-user-session-
collection-500k). Columns include `Query`; we derive counts by AGGREGATION
(COUNT(*) per normalized query) — exactly the assignment's "derive counts if
the dataset has none". The query-popularity distribution is naturally Zipf,
which lets the performance report demonstrate the Pareto caching argument.

Usage (from project root, venv active):
  python -m scripts.load_dataset --file files/user-ct-test-collection-01.txt
  python -m scripts.load_dataset --synthetic 120000        # no download needed

The loader auto-detects delimiter (AOL files are TAB-separated) and the Query
column from the header.
"""
import argparse
import asyncio
import csv
import glob
import gzip
import sys
from collections import Counter
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT))

from backend import store  # noqa: E402


def _open(path: str):
    """Open .gz transparently."""
    if path.endswith(".gz"):
        return gzip.open(path, "rt", encoding="utf-8", errors="ignore", newline="")
    return open(path, "r", encoding="utf-8", errors="ignore", newline="")


def _aggregate_into(path: str, counter: Counter) -> None:
    with _open(path) as f:
        sample = f.read(4096)
        f.seek(0)
        delim = "\t" if sample.count("\t") >= sample.count(",") else ","
        reader = csv.reader(f, delimiter=delim)
        header = next(reader, None)
        if not header:
            return
        lowered = [h.strip().lower() for h in header]
        try:
            qi = lowered.index("query")
        except ValueError:
            qi = 1 if len(header) > 1 else 0  # AOL: AnonID, Query, ...
        n = 0
        for row in reader:
            if len(row) <= qi:
                continue
            q = row[qi].strip().lower()
            if not q or q == "-":
                continue
            counter[q] += 1
            n += 1
        print(f"  {Path(path).name}: {n:,} query rows (delim={delim!r}, col={qi})")


def aggregate_paths(paths: list[str], min_count: int) -> list[tuple[str, int]]:
    counter: Counter[str] = Counter()
    for p in paths:
        _aggregate_into(p, counter)
    return [(q, c) for q, c in counter.items() if c >= min_count]


def write_tsv(rows: list[tuple[str, int]], path: str) -> None:
    """Persist aggregated (query, count) so the expensive aggregation is reusable."""
    with open(path, "w", encoding="utf-8") as f:
        for q, c in rows:
            f.write(f"{q}\t{c}\n")


def read_agg(path: str, top: int | None, min_count: int) -> list[tuple[str, int]]:
    """Stream a pre-aggregated query<TAB>count file (low memory). Optionally keep
    only the top-N by count."""
    rows: list[tuple[str, int]] = []
    with open(path, "r", encoding="utf-8", errors="ignore") as f:
        for line in f:
            parts = line.rstrip("\n").split("\t")
            if len(parts) != 2:
                continue
            q, c = parts[0], int(parts[1])
            if q and c >= min_count:
                rows.append((q, c))
    if top:
        rows.sort(key=lambda x: x[1], reverse=True)
        rows = rows[:top]
    return rows


def synthetic(n: int) -> list[tuple[str, int]]:
    """Generate ~n unique queries with Zipf-distributed counts (no download)."""
    heads = ["how to", "best", "buy", "cheap", "free", "download", "what is",
             "iphone", "samsung", "java", "python", "amazon", "google", "weather",
             "news", "movie", "song", "recipe", "near me", "online", "review"]
    tails = ["tutorial", "price", "2026", "review", "online", "near me", "for sale",
             "guide", "vs", "app", "login", "meaning", "today", "free", "pro max",
             "case", "charger", "stock", "results", "live", "download", "lyrics"]
    mids = ["", "best ", "new ", "cheap ", "top ", "the "]
    out: dict[str, int] = {}
    rank = 1
    i = 0
    while len(out) < n:
        h = heads[i % len(heads)]
        m = mids[(i // len(heads)) % len(mids)]
        t = tails[(i // (len(heads) * len(mids))) % len(tails)]
        suffix = f" {i // 5000}" if i >= 5000 else ""
        q = f"{h} {m}{t}{suffix}".strip()
        if q not in out:
            count = max(1, int(10_000_000 / (rank ** 1.1)))  # Zipf-ish
            out[q] = count
            rank += 1
        i += 1
    return list(out.items())


async def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--file", help="path to a single AOL query log (.txt or .txt.gz)")
    ap.add_argument("--dir", help="directory of AOL files (loads all *.txt and *.txt.gz)")
    ap.add_argument("--synthetic", type=int, help="generate N synthetic queries instead")
    ap.add_argument("--agg-file", help="load a pre-aggregated query<TAB>count file (low memory)")
    ap.add_argument("--out", help="aggregate to this TSV file and exit (no DB write)")
    ap.add_argument("--top", type=int, help="keep only the top-N queries by count")
    ap.add_argument("--min-count", type=int, default=1, help="drop queries below this count")
    ap.add_argument("--no-truncate", action="store_true", help="append instead of replace")
    args = ap.parse_args()

    if not any([args.file, args.dir, args.synthetic, args.agg_file]):
        raise SystemExit("provide --file, --dir, --agg-file, or --synthetic")

    print("aggregating...")
    if args.synthetic:
        rows = synthetic(args.synthetic)
    elif args.agg_file:
        rows = read_agg(args.agg_file, args.top, args.min_count)
    else:
        if args.dir:
            paths = sorted(glob.glob(f"{args.dir}/*.txt.gz") + glob.glob(f"{args.dir}/*.txt"))
            if not paths:
                raise SystemExit(f"no .txt/.txt.gz files in {args.dir}")
        else:
            paths = [args.file]
        rows = aggregate_paths(paths, args.min_count)
        if args.top:
            rows.sort(key=lambda x: x[1], reverse=True)
            rows = rows[: args.top]

    # aggregate-only mode: persist to TSV and stop (DB not required)
    if args.out:
        write_tsv(rows, args.out)
        print(f"wrote {len(rows):,} rows to {args.out}")
        return
    print(f"{len(rows):,} distinct queries to load")
    if len(rows) < 100_000:
        print(f"WARNING: only {len(rows):,} queries (< 100k assignment minimum)")

    await store.init_pool()
    try:
        await store.init_schema()
        if not args.no_truncate:
            await store.truncate()
        await store.bulk_load(rows)
        total = await store.count_rows()
        print(f"done. rows in DB: {total:,}")
    finally:
        await store.close_pool()


if __name__ == "__main__":
    asyncio.run(main())
